draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,43 +60,231 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+in the extension, malicious leader cannot change the task configuration to
+mislead helpers to reduce privacy guarantee, for e.g. by reducing
+`min_batch_size`.
+
+This extension affects upload, aggregate and collect sub-protocols.
 
 # Conventions and Definitions
 
 {::boilerplate bcp14-tagged}
 
+Similar to {{?DAP=I-D.draft-ietf-ppm-dap-01}}, this document uses the verbs
+"abort" and "alert with `[some error message]`" to describe how protocol
+participants react to various error conditions.
+
+The terms used follow those described in {{?DAP=I-D.draft-ietf-ppm-dap-01}}. The
+following new terms are used:
+
+Task provision:
+: The process of creating a DAP task.
+
+Task configuration:
+: The non-secret parameters required to create a task in task proviosion.
+
+Task author:
+: The entity that defines the parameters of a task.
 
-# The "task_prov" Extension
+
+# The "task-prov" Extension
 
 > NOTE(cjpatton) It will be useful to think of a good name for this. Something
-> succinct and descriptive is ideal. `task_prov` is fine, but we can maybe do
+> succinct and descriptive is ideal. `task-prov` is fine, but we can maybe do
 > better.
 
 A new extension is defined:
 
 ~~~
 enum {
-   task_prov(0xff00), (65535)
+    task-prov(0xff00),
+    (65535)
 } ExtensionType;
 ~~~
+> NOTE(shan) TLS uses underscore `_` to define codepoints, which is different
+> from DAP's convension of using hyphen `-`, which one to follow?
+> See [#314](ietf-wg-ppm/draft-ietf-ppm-dap#314) for discussion.
 
 When the Client includes this extension with its report, the body of the
 extension is structured as follows:
 
 ~~~
-TODO(wangshan) Specify the structure of the extension.
+/* Definition of all parameters in extension_data for one task */
+struct {
+    /* Info specific for a task. */
+    opaque task_info<1..2^8-1>;
+
+    /* A list of URLs relative to which an aggregator's API endpoints */
+    /* can be found. */
+    Url aggregator_endpoints<1..2^16-1>;
+
+    /* This determines the query type for batch selection and the */
+    /* properties that all batches for this task must have. */
+    QueryConfig query_config;
+
+    /* The maximum number of times a batch of reports may be queried */
+    /* by the Collector. */
+    uint16 max_batch_lifetime;
+
+    /* Time up to which clients are allowed to upload to this task. */
+    /* See https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/304 */
+    Time task_expiration;
+
+    /* A unique identifier for the VDAF instance used for the task, */
+    /* including the type of measurement associated with the task. */
+    VdafType vdaf_type;
+
+    /* Additional parameters relevant for the vdaf_type */
+    opaque vdaf_data<1..2^16-1>;
+} TaskConfig;
+
+/* Defined in DAP core protocol. */
+enum {
+    reserved(0), // Reserved for testing purposes
+    time-interval(1),
+    fixed-size(2),
+    (65535)
+} QueryType;
+
+struct {
+    uint32 max_batch_size;
+} FixedSizeQueryConfig;
```

Comment:
wouldn't it be more clear if we have an explicitly type for it? for e.g. if we have another field required by FixedSizeQueryConfig in the future

Reply:
It's just my opinion, but I don't think this is more clear, since you now have to look in two places to understand the information conveyed by QueryConfig. However, defining new structs is sometimes appropriate. For example:
1.  when either you need to serialize the same data in multiple places in the protocol
2. when there are a large number of parameters in the struct.

Reply:
I'll change this since I don't know what else should be include in FixedSizeQueryConfig yet.